integrated_channels/canvas/client.py

```diff
@@ -44,7 +47,78 @@ def update_content_metadata(self, serialized_data):
     def delete_content_metadata(self, serialized_data):
         pass
 
-    def _get_oauth_access_token(self, client_id, client_secret):
+    def _post(self, url, data, scope):
+        """
+        Make a POST request using the session object to a Canvas endpoint.
+
+        Args:
+            url (str): The url to send a POST request to.
+            data (str): The json encoded payload to POST.
+        """
+        self._create_session()
+        response = self.session.post(url, data=data)
+        return response.status_code, response.text
+
+    def _delete(self, url, data, scope):
         """
-        TODO: get oauth token for canvas api access
+        Make a DELETE request using the session object to a Canvas endpoint.
+
+        Args:
+            url (str): The url to send a DELETE request to.
+            data (str): The json encoded payload to DELETE.
+        """
+        self._create_session()
+        response = self.session.delete(url, data=data)
+        return response.status_code, response.text
+
+    def _create_session(self):
         """
+        Instantiate a new session object for use in connecting with Canvas. Each enterprise customer
+        connecting to Canvas should have a single client session.
+        """
+        if self.session is None:
+            # Create a new session with a valid token
+            if self.session:
+                self.session.close()
+            oauth_access_token = self._get_oauth_access_token(
+                self.enterprise_configuration.client_id,
+                self.enterprise_configuration.secret,
+            )
+            session = requests.Session()
+            session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)
+            session.headers['content-type'] = 'application/json'
+            self.session = session
+
+    def _get_oauth_access_token(self, client_id, client_secret):
+        """Uses the client id, secret and refresh token to request the user's auth token from Canvas.
+
+        Args:
+            client_id (str): API client ID
+            client_secret (str): API client secret
+
+        Returns:
+            access_token (str): the OAuth access token to access the Canvas API as the user
+        Raises:
+            HTTPError: If we received a failure response code from Canvas.
+            RequestException: If an unexpected response format was received that we could not parse.
+        """
+        auth_token_url = urljoin(
+            self.enterprise_configuration.canvas_base_url,
+            self.config.oauth_token_auth_path,
+        )
+        auth_token_params = {
+            'grant_type': 'refresh_token',
+            'client_id': client_id,
+            'client_secret': client_secret,
+            'state': self.enterprise_configuration.enterprise_customer.uuid,
+            'refresh_token': self.enterprise_configuration.refresh_token,
```

Comment:
yeah it may be worth doc'ing somewhere in the class level doc maybe? that the value of refresh_token here should already be set in the enterpriseccustomerconfig, and if it isn't we should raise an exception? that is a different failure mode than, say, the code expiring

Reply:
Docing the whole process I think will be beneficial for ECS as well as any other engineer who may stumble across oauth 👍 . 

I could check for the presence of the refresh token, but the canvas API has some pretty straightforward error responses if you try and request without one. `{"error":"invalid_request","error_description":"You must provide the refresh_token parameter when using the refresh_token grant type"}`. My intuition is that either approach is viable here, but can default to your preferences if you think there's a standard we should adhere to.

Reply:
yeah ok if this output will echo into the logs when we get this that's great!

This maybe leads into some testing thoughts:

I was mainly thinking about explicitly locking down the various outcomes that can happen, ideally in small unit test cases

for instance can we have test cases unit level where we can get:

1. expired code hence cannot get refresh_token?
2. client_id is incorrect or sth like that and we get an invalid or auth response code? not sure what happens in that case
3. anything else you can think of that can go wrong, there may be more that I cannot think of right now

A really good way to do such things is to have explicit exceptions named appropriately and then unit tests can just inject the fails and check for the right exception, just a thought if not too much work

We should basically assume things will go wrong occasionally in this process (such as a temporary glitch in the server, and a 500 maybe for example), so in the very least maybe test at least one or two of the salient negative cases would be great

Reply:
chatted with Alex. Alex, you can safely ignore my notes on 'code' expiring since that code will be used to call back the callback_uri so there is no notino of them expiring on our side. but let's find out about refresh_token itself (if they expire and how deal with that). But yeah the general thoughts on error handling are still pertinent I hope

Reply:
> 1. expired code hence cannot get refresh_token?

Discussed offline, the client/client session shouldn't save any reference to the code and to my knowledge refresh tokens don't expire (currently a 3 day old token isn't expired and nothing in their docs that I've found indicates an expiration **on the refresh token**. If the refresh token is invalid at the time of the request, we get a similar response `{"error":"invalid_request","error_description":"refresh_token not found"}`

> 2. client_id is incorrect or sth like that and we get an invalid or auth response code? not sure what happens in  that case

In the same line of thinking, Canvas responds with `{"error":"invalid_client","error_description":"unknown client"}` if you have an invalid client id, which I think is sufficient to let any dev know what's going wrong in the logs. 

Because of these canvas exception responses, it's difficult to think of situations that are testable for the client. Open to suggestions but going to take a moment and try and think of potential edge cases.

Reply:
no worries, we can test manually too when we are testing this. I am happy to lend hands in that regard too if we want to do that even after merge is okay just to see what things look like. what happens often is once we ship to prod we find it hard to diagnose without the right exceptions, let's just do the best we can at this stage of the work, we are doing great!